### ISSUE TITLE
How would I visualize detection outputs during training?

### ISSUE THREAD

**Issue by zlenyk** [open-mmlab/mmdetection#7093]
Hi, how would I go about implementing visualization of detection during training?
In ideal world, that would be a new hook that would run after each X epochs, take some images from validation set and save to file it's visualization.
I'm finding it impossible to implement for few reasons:
- I don't have access to data loaders from hooks
- I don't have access to test pipelines from hooks

Things that I considered:
- using mmdet.apis.inference_detector: this doesn't work, because it takes as argument some weird structure that is result of init_detector (for example config is added to it), so using it without running init_detector is hard.  However I can't run init_detector because that would require to load weights from checkpoint, which would make me save the checkpoint first - generally quite annoying workflow.
- using runner.model from the hook directly, however there's no way to access data loader so it's not possible to format images correctly.

I'm quite stuck with this. Perfect solution would be to:
1. Access train/val dataloader from hook
2. Run inference of runner.model
3. Run runner.model.show_results to save the result

Is there a recommended way or even some workaround to do it?

Please note that I'm aware that I can do it after the training via accessing model checkpoint, however it'd be nice to have it also as training hook.

Thank you!

**Comment by zlenyk** [comment]
Pinging this thread. Please let me know if this is not the right place to ask this kind of questions.

**Comment by Chop1** [comment]
Hello,

I'm facing the same problem as you. While creating custom hook, I dont see how to access the existing writter, such as the tensorflow writter.

You can still write log using this workaround, but it's messy..

go to your detector and update your train_step (or val step if you want log from validation) to pass your stuff
https://github.com/open-mmlab/mmdetection/blob/7a9bc498d5cc972171ec4f7332afcd70bb50e60e/mmdet/models/detectors/base.py#L221

```python
    def train_step(self, data, optimizer):

        losses = self(**data)
        loss, log_vars = self._parse_losses(losses)

        
        # Here I access the tensors I want to log
        l_tensor_attention = []
        for att_blocks in self._modules["detector"].neck.decoder.att_blocks:
            for attention_gates in att_blocks.attention_gates:
                l_tensor_attention.append(attention_gates.psi.layers.activate)

        tensor_input_img = self._modules["detector"].backbone.conv1

        # add them to the output, wich is accessed by hooks
        outputs = dict(
            loss=loss, log_vars=log_vars, num_samples=len(data['img_metas']),
            l_tensor_attention=l_tensor_attention, tensor_input_img=tensor_input_img)

        return outputs
```

then you go to your logger hook (I use an existing hook, because I dont know how to access the logger from a new hook I would write)
https://github.com/open-mmlab/mmcv/blob/b83bdb0cbee98c5e73ff193042a832381565d313/mmcv/runner/hooks/logger/tensorboard.py#L59

```python
def log(self, runner):

        def viz_input_hook(inst, inp, out):
            if (self.get_iter(runner) % 250) == 0:

                invTrans = transforms.Compose([ transforms.Normalize(mean = [ 0., 0., 0. ],
                                                     std = [ 1/58.395, 1/57.12, 1/57.375 ]),
                                transforms.Normalize(mean = [ -123.675, -116.28, -103.53 ],
                                                     std = [ 1., 1., 1. ]),
                               ])

                input_img = inp[0].detach().clone()
                input_img = invTrans(input_img)

                grid_img = utils.make_grid(input_img, nrow=1, normalize=True, value_range=(0, 255))
                plt.figure(figsize = (3, 1*out.shape[0]+2))
                plt.imshow(grid_img.detach().permute(1, 2, 0).cpu().numpy()[:, :, :], vmin=0, vmax=255, interpolation='nearest')
                plt.tight_layout()
                plt.axis('off')

                self.writer.add_figure(f"image/input", plt.gcf(), self.get_iter(runner))
        

        def viz_activation_hook(inst, inp, out, level):

            if (self.get_iter(runner) % 250) == 0:

                grid_img = utils.make_grid(out, nrow=1)
                grid_numpy = grid_img.detach().permute(1, 2, 0).cpu().numpy()

                plt.figure(figsize = (3, 1*out.shape[0]+2))

                averaged_attention = np.median(grid_numpy, keepdims=True, axis=2)
                plt.imshow(averaged_attention, vmin=0, vmax=1, cmap="jet", interpolation='nearest')

                plt.tight_layout()
                plt.axis('off')

                self.writer.add_figure(f"image/{level}_color_{repr(inst)}", plt.gcf(), self.get_iter(runner))

        # terrible hack to register the hook once
        if (self.get_iter(runner) == 100):
            runner.outputs.get('tensor_input_img').register_forward_hook(viz_input_hook)
            for i, layer in enumerate(runner.outputs.get('l_tensor_attention')):
                layer.register_forward_hook(partial(viz_activation_hook, level=i))
        


        for tag, val in tags.items():
            if isinstance(val, str):
                self.writer.add_text(tag, val, self.get_iter(runner))
            else:
                self.writer.add_scalar(tag, val, self.get_iter(runner))
```

and voila, terrible code but it works. If you have a more clean solution I would gladly know about it.
